tests/unit/scripting.tcl

```diff
@@ -1336,3 +1216,155 @@ start_server {tags {"scripting needs:debug"}} {
     r debug set-disable-deny-scripts 0
 }
 } ;# foreach is_eval
+
+
+# Scripting "shebang" notation tests
+start_server {tags {"scripting"}} {
+    test "Shebang support for lua engine" {
+        catch {
+            r eval {#!not-lua
+                return 1
+            } 0
+        } e
+        assert_match {*Unexpected engine in script shebang*} $e
+
+        assert_equal [r eval {#!lua
+            return 1
+        } 0] 1
+    }
+
+    test "Unknown shebang option" {
+        catch {
+            r eval {#!lua badger=data
+                return 1
+            } 0
+        } e
+        assert_match {*Unknown lua shebang option*} $e
+    }
+
+    test "Unknown shebang flag" {
+        catch {
+            r eval {#!lua flags=allow-oom,what?
+                return 1
+            } 0
+        } e
+        assert_match {*Unexpected flag in script shebang*} $e
+    }
+
+    test "allow-oom shebang flag" {
+        r set x 123
+    
+        r config set maxmemory 1
+
+        # Fail to execute deny-oom command in OOM condition (backwards compatibility mode without flags)
+        assert_error {ERR Error running script *OOM command not allowed when used memory > 'maxmemory'.} {
+            r eval {
+                redis.call('set','x',1)
+                return 1
+            } 1 x
+        }
+        # Can execute non deny-oom commands in OOM condition (backwards compatibility mode without flags)
+        assert_equal [
+            r eval {
+                return redis.call('get','x')
+            } 1 x
+        ] {123}
+
+        # Fail to execute regardless of script content when we use default flags in OOM condition
+        assert_error {OOM allow-oom flag is not set on the script, can not run it when used memory > 'maxmemory'} {
+            r eval {#!lua flags=
+                return 1
+            } 0
+        }
+
+        assert_equal [
+            r eval {#!lua flags=allow-oom
+                redis.call('set','x',1)
+                return 1
+            } 0
+        ] 1
+
+        r config set maxmemory 0
+    }
+
+    test "no-writes shebang flag" {
+        assert_error {ERR Error running script *Write commands are not allowed from read-only scripts.} {
+            r eval {#!lua flags=no-writes
+                redis.call('set','x',1)
+                return 1
+            } 1 x
+        }
+    }
+    
+    start_server {tags {"external:skip"}} {
+        r -1 set x "some value"
+        test "no-writes shebang flag on replica" {
+            r slaveof [srv -1 host] [srv -1 port]
```

Comment:
in new code, we try to avoid using "slave", less changes if we some day remove it.
adding this petty comment just to note to myself that i reviewed this batch of commits.
```suggestion
            r replicaof [srv -1 host] [srv -1 port]
```